packages/pwa/app/utils/url.js

```diff
@@ -102,16 +103,23 @@ export const productUrlBuilder = (product) => encodeURI(`/product/${product.id}`
 export const searchUrlBuilder = (searchTerm) => `/search?q=${searchTerm}`
 
 /**
- * Returns a URL using the new locale.
+ * Returns a URL pathname using the new locale.
  *
- * @param previousLocale
- * @param newLocale
+ * @param {string} newLocale
+ * @param {Object} [options]
+ * @param {Object} options.location - original location (similar value to `window.location`)
  * @returns {string}
  */
-export const buildUrlLocale = (previousLocale, newLocale) => {
+export const buildUrlLocale = (newLocale, options = {}) => {
+    const location = options.location
+        ? options.location
+        : typeof window !== 'undefined'
+        ? window.location
+        : useLocation()
```

Comment:
This change allows for:
- passing in an optional `location` object, which is great for testing purpose
- calling `useLocation` hook for working SSR rendering

Reply:
IMO,  if we were to favor the `options.location` just for the testing purpose, I think it is not good enough to use it. Maybe we should mock the location in the unit test instead of writing the function to serve the testing purpose.

Reply:
I think it's more flexible if the function allows you to pass in a location. It's not only for testing purpose. 

This is similar to discussions about having pure functions without side effect :) There are other people out there who prefer _not_ to mock things in their tests if possible. To them, mocking things is like code smell. 

For me, I'm not that strict. For _utility_ functions, I still think, though, that it would be better if they're more pure functions.